ouroboros-consensus-test/bench-mempool/Bench/Consensus/Mempool/Scenario.hs

```diff
@@ -0,0 +1,227 @@
+    in
+        assert (Set.null fullConsumed)
+      $ modifyBackingTokens (const $ Set.toList fullProduced)
+
+modifyChangelogBlocks :: MonadState St m => ([TestBlock] -> [TestBlock]) -> m ()
+modifyChangelogBlocks f = modifyScenario (\sc -> sc {
+      scChangelogBlocks = f $ scChangelogBlocks sc
+    })
+
+putChangelogBlocks :: MonadState St m => [TestBlock] -> m ()
+putChangelogBlocks blks = modifyChangelogBlocks (const blks)
+
+modifyMempoolCommands ::
+     MonadState St m
+  => ([MempoolCmd TestBlock] -> [MempoolCmd TestBlock]) -> m ()
+modifyMempoolCommands f = modifyScenario (\sc -> sc {
+      scMempoolCommands = f $ scMempoolCommands sc
+    })
+
+putMempoolCommands :: MonadState St m => [MempoolCmd TestBlock] -> m ()
+putMempoolCommands cmds = modifyMempoolCommands (const cmds)
+
+{-------------------------------------------------------------------------------
+  Scenario builder: utilities
+-------------------------------------------------------------------------------}
+
+-- | Monadic equivalent to 'iterate'. The 'Int' parameter ensures termination.
+--
+-- Note: based on 'iterateM' from @monad-extras-0.6.0@.
+iterateM :: forall m a. Monad m => Int -> (a -> m a) -> m a -> m [a]
+iterateM n0 f xm0
+  | n0 <= 0   = error "iterateM: n <= 0"
+  | otherwise = go n0 xm0
+  where
+    go :: Int -> m a -> m [a]
+    go 0 _  = pure []
+    go n xm = do
+      x <- xm
+      (x:) <$> go (n-1) (f x)
+
+{-------------------------------------------------------------------------------
+  Scenario builder: primitives
+-------------------------------------------------------------------------------}
+
+newToken :: MonadState St m => m Token
+newToken = do
+  nxtTok <- gets stNextToken
+  modify (\st -> st { stNextToken = nxtTok + 1 })
+  pure nxtTok
+
+-- | @'txN' txsTC n@ creates a 'Tx' that consumes all 'Token's
+-- produced by the 'Tx's in @txsTC@, and produces @n@ new 'Token's.
+txN :: MonadState St m => [Tx] -> Int -> m Tx
+txN txsTC n = do
+  let toksTC = Set.toList $ foldMap produced txsTC
+  toksTP <- replicateM n newToken
+  pure $ mkTx toksTC toksTP
+
+-- | @'tx1' txTC n@ creates a 'Tx' that consumes all 'Token's
+-- produced by @txTC@, and produces @n@ new 'Token's.
+_tx1 :: MonadState St m => Tx -> Int -> m Tx
+_tx1 txTC = txN [txTC]
+
+-- | @'stxN' txsTC@ creates a 'Tx' that consumes all 'Token's produced by the
+-- 'Tx's in @txsTC@, and produces only @1@ new 'Token'.
+_stxN :: MonadState St m => [Tx] -> m Tx
+_stxN txTC = txN txTC 1
+
+-- | @'stx1' txTC@ creates a 'Tx' that consumes all 'Token's produced
+-- by @txTC@, and produces only @1@ new 'Token'.
+stx1 :: MonadState St m => Tx -> m Tx
+stx1 txTC = txN [txTC] 1
+
+-- | @'gtxN' txTC@ creates a 'Tx' that consumes no 'Token's, and produces @n@
+-- new 'Token's.
+_gtxN :: MonadState St m => Int -> m Tx
+_gtxN = txN []
+
+-- | @'gtx1' txTC@ creates a 'Tx' that consumes no 'Token's, and produces only
+-- @1@ new 'Token's.
+gtx1 :: MonadState St m => m Tx
+gtx1 = txN [] 1
```

Comment:
I'm unsure how to give these functions nice, descriptive names without making them very long

Reply:
See https://github.com/input-output-hk/ouroboros-network/pull/4476#discussion_r1165394490